### ISSUE TITLE
PossiblyNullPropertyAssignment false positive

### ISSUE THREAD

**Issue by patrickkusebauch** [vimeo/psalm#3200]
Just a bit of context before code.

 - I have a (Nette) Presenter class.
 - In startup/constructor, I ask the domain for an entity by ID
 - the domain can respond `null` in case there is no entity of that ID.
 - I immediately check for `null` condition - in case it is null, I call a function that throws an uncaught Exception
- after this check (no chance therefore of it being `null`) I assign to typed class property
- this causes psalm error and it should not

Simplified example: https://psalm.dev/r/014271d6e6

**Comment by weirdan** [comment]
If your exception-throwing function never returns (in a normal way) annotate it as such: https://psalm.dev/r/587f0e8602 and all Psalm issues go away.

Although I think Psalm should still complain about potentially uninitialized property, because public, non-final method may be overridden in descendants: https://psalm.dev/r/4f967f6930, https://3v4l.org/juGjD

**Comment by patrickkusebauch** [comment]
@weirdan  `public function redirect` is a library function, I cannot change its annotation. It is actually a parent of the presenter class normally - in real code: https://github.com/nette/application/blob/master/src/Application/UI/Component.php#L281

I have tried to simplify the issue.

**Comment by weirdan** [comment]
Well, you have a number of options:

1. Create a stub for that base class and annotate it with `never-returns` (see [docs](https://psalm.dev/docs/running_psalm/configuration/#ltstubsgt). This is preferred option.
2. Add a (never reachable) `return`, `throw` or `exit()` after the call to the function that never returns. Of all of those I would probably prefer `return`.
3. Add an assertion on the alternative path, like this: https://psalm.dev/r/e229cd533d

Basically Psalm doesn't know the method execution terminates in that `if`. And I don't think it could infer it, especially if the call to `terminate()` (which throws the exception) is made several calls stack frames deeper.
